lte/gateway/python/magma/pipelined/app/apn.py

```diff
@@ -0,0 +1,284 @@
+        """
+        Install the default flows on datapath connect event.
+
+        Args:
+            datapath: ryu datapath struct
+        """
+        self._datapath = datapath
+        # In case wee need to clean all existing  buggy / orphaned flows 
+        # before start the controller
+        if self._clean_start:
+          self.delete_all_flows(datapath)
+
+    @set_ev_cls(ofp_event.EventOFPBarrierReply, MAIN_DISPATCHER)
+    def _handle_barrier(self, ev):
+        self._msg_hub.handle_barrier(ev)
+
+    @set_ev_cls(ofp_event.EventOFPErrorMsg, MAIN_DISPATCHER)
+    def _handle_error(self, ev):
+        self._msg_hub.handle_error(ev)
+
+    def _wait_for_flow_responses(self, imsi, flow_list, chan):
+        def fail(err):
+            self.logger.error(
+                "Failed to install flow for subscriber %s: %s",
+                imsi, err)
+            return False
+
+        for _ in range(len(flow_list)):
+            try:
+                result = chan.get()
+            except MsgChannel.Timeout:
+                return fail("No response from OVS")
+            if not result.ok():
+                return fail(result.exception())
+        return True
+
+    def cleanup_on_disconnect(self, datapath):
+        """
+        Cleanup flows on datapath disconnect event.
+
+        Args:
+            datapath: ryu datapath struct
+        """
+        self.delete_all_flows(datapath)
+
+    def delete_all_flows(self, datapath):
+        """Delete all flows which set APN register"""
+        flows.delete_all_flows_from_table(datapath, self.tbl_num)
+        flows.delete_all_flows_from_table(datapath, self.apn_tagging_scratch)
+
+    def add_apn_flow_for_ue(self, imsi, ue_ip_addr, apn):
+        """ 
+            Add flow which match all IN traffic with specified UE_IP and set 
+            APN hash in to register.
+            
+            Args:
+                imsi: user's IMSI
+                ue_ip_addr: ip addr allocated for the UE in scope of connection 
+                    to specific APN
+                apn: APN UE is connected to with specified IP addr
+        """
+        #TODO(119vik): same IP is reused for several bearers connected to the 
+        # same APN - take care about duplications
+        parser = self._datapath.ofproto_parser
+        
+        encoded_apn = encode_apn(apn)
+        apn_tagging_actions = self._get_apn_tagging_actions(parser, encoded_apn)
+        flow_adds = []
+
+        # Prepare matchers for both UpLink and DownLink traffic
+        outbound_match = MagmaMatch(
+            direction=Direction.OUT, 
+            ipv4_src=ue_ip_addr, 
+            eth_type=ether.ETH_TYPE_IP)
+        inbound_match = MagmaMatch(
+            direction=Direction.IN, 
+            ipv4_dst=ue_ip_addr, 
+            eth_type=ether.ETH_TYPE_IP)
+        
+        # Generate OF Add Flow messages for Uplink and DownLink traffic
+        flow_adds += self._get_flows_for_traffic_direction(
```

Comment:
Passing a generator seems too complex. You should just do singular flow_adds += rule (pull out the generator and use it directly)

Reply:
reworked a bit.